chat-plugins/mafia.js

```diff
@@ -852,6 +915,194 @@ class MafiaTracker extends Rooms.RoomGame {
+			selection[1] = player.IDEA.choices.splice(roleIndex, 1)[0];
+		} else {
+			selection[1] = '';
+		}
+
+		if (player.IDEA.picks[selection[0]]) {
+			buf += `You have deselected ${player.IDEA.picks[selection[0]]}. `;
+			player.IDEA.choices.push(player.IDEA.picks[selection[0]]);
+		}
+
+		if (player.IDEA.picks[selection[0]] && !selection[1]) {
+			this.IDEA.waitingPick.push(player.userid);
+		} else if (!player.IDEA.picks[selection[0]] && selection[1]) {
+			this.IDEA.waitingPick.splice(this.IDEA.waitingPick.indexOf(player.userid), 1);
+		}
+
+		player.IDEA.picks[selection[0]] = selection[1];
+		if (selection[1]) buf += `You have selected ${selection[0]}: ${selection[1]}.`;
+		player.updateHtmlRoom();
+		if (!this.IDEA.waitingPick.length) {
+			if (this.IDEA.timer) clearTimeout(this.IDEA.timer);
+			this.ideaFinalizePicks();
+			return;
+		}
+		return user.sendTo(this.room, buf);
+	}
+
+	ideaFinalizePicks() {
+		if (!this.IDEA || !this.IDEA.data) return this.sendRoom(`Tried to finalize IDEA picks with no IDEA module running, please report this to a mod.`);
+		let randed = [];
+		for (const p in this.players) {
+			const player = this.players[p];
+			if (!player.IDEA) return this.sendRoom(`Trying to pick an IDEA role with no player IDEA object, user: ${player.userid}. Please report this to a mod.`);
+			let randPicked = false;
+			let role = [];
+			for (const choice of this.IDEA.data.picks) {
+				if (!player.IDEA.picks[choice]) {
+					randPicked = true;
+					player.IDEA.picks[choice] = player.IDEA.choices.shift();
+					const user = Users(player.userid);
+					if (user) user.sendTo(this.room, `You were randomly assigned ${choice}: ${player.IDEA.picks[choice]}`);
+				}
+				role.push(`${choice}: ${player.IDEA.picks[choice]}`);
+			}
+			if (randPicked) randed.push(p);
+			// if there's only one option, it's their role, parse it properly
+			let roleName = '';
+			if (this.IDEA.data.picks.length === 1) {
+				const role = parseRole(player.IDEA.picks[this.IDEA.data.picks[0]]);
+				player.role = role.role;
+				if (role.problems.length) this.sendRoom(`Problems found when parsing IDEA role ${player.IDEA.picks[this.IDEA.data.picks[0]]}. Please report this to a mod.`);
+			} else {
+				roleName = role.join('; ');
+				player.role = {
+					name: roleName,
+					safeName: Chat.escapeHTML(roleName),
+					id: toId(roleName),
+					alignment: 'solo',
+					memo: [`(Your role was set from an IDEA.)`],
+					image: '',
+				};
+			}
+		}
+		this.IDEA.discardsHtml = `<b>Discards:</b><br />`;
+		for (const p of Object.keys(this.players).sort()) {
+			// @ts-ignore guaranteed above
+			this.IDEA.discardsHtml += `<b>${this.players[p].safeName}:</b> ${this.players[p].IDEA.choices.join(', ')}<br />`;
+		}
+
+		this.phase = 'IDEAlocked';
+		if (randed.length) this.sendRoom(`${randed.join(', ')} didn't pick.`, {declare: true});
+		this.sendRoom(`IDEA picks are locked!`, {declare: true});
```

Comment:
I would be more descriptive here. eg: `${randed.join(', ')} did not pick a role in time and were randomly assigned one.`